parse_test.go

```diff
@@ -116,3 +116,63 @@ func TestParseFile(t *testing.T) {
 	}
 	assert.Equal(t, expected, cmds)
 }
+
+func TestParseReaderHeredocs(t *testing.T) {
+	dockerfile := `RUN <<EOF
+source $HOME/.bashrc && echo $HOME
+echo "Hello" >> /hello
+echo "World!" >> /hello
+EOF
+`
+	cmds, err := ParseReader(bytes.NewBufferString(dockerfile))
+	assert.Nil(t, err)
+	expected := []Command{
+		Command{
+			Cmd:       "RUN",
+			Original:  dockerfile,
+			StartLine: 1,
+			EndLine:   5,
+			Flags:     []string{},
+			Value:     []string{"source $HOME/.bashrc && echo $HOME\necho \"Hello\" >> /hello\necho \"World!\" >> /hello\n"},
+			Heredocs: []Heredoc{
+				Heredoc{
+					Name:           "EOF",
+					FileDescriptor: 0,
+					Content:        "source $HOME/.bashrc && echo $HOME\necho \"Hello\" >> /hello\necho \"World!\" >> /hello\n"},
+			},
+		},
+	}
+	assert.Equal(t, expected, cmds)
+}
+
+func TestParseReaderHeredocsMultiple(t *testing.T) {
+	dockerfile := `COPY <<FILE1 <<FILE2 /dest
+content 1
+FILE1
+content 2
+FILE2
+`
+	cmds, err := ParseReader(bytes.NewBufferString(dockerfile))
+	assert.Nil(t, err)
+	expected := []Command{
+		Command{
+			Cmd:       "COPY",
+			Original:  dockerfile,
+			StartLine: 1,
+			EndLine:   5,
+			Flags:     []string{},
+			Value:     []string{"content 1\n", "content 2\n"},
```

Comment:
where did `/dest` go? this doesn't seem right

Reply:
I totally agree, but unfortunately that is how buildkit is parsing that case. If you look at the sample test case they have at https://github.com/moby/buildkit/blob/master/frontend/dockerfile/parser/parser_heredoc_test.go line 67, then the expected value at 204, the /dest does disappear. I am not sure why they are doing this, but since we are not trying to improve on their parser but just support what they do, this is the way our test case is behaving as well. At most, we could think about putting it in some manner in the Value field, which I was kinda doing before, but changes in the heredoc substructure should depend on changes they make.

Reply:
the whole point of this is to allow one to get the whole structural syntax by improving on the parser

Reply:
Not really. The whole point is to bring parity with the current version of the docker parser in buildkit, which has heredocs support. Since this library is just a wrapper around that, the best we can do is to wrap around what that parser is doing. If that parser has bugs or shortcomings, the proper place to fix it is over there.

Reply:
That being said, I realized that the contents of the Value field now for the heredocs do not make much sense, given the new Heredocs structure and the fact that the Heredocs contents are now in there. I pushed a new change where I modified the Value field to go back to how it works for all command, which for the Heredocs case means that it will parse the first line of the command, including the names of the heredocs and anything else in that line, like the "/dest" param in that line. The Heredocs substructure will work as the attachment with the contents of the heredocs, which is how buildkit does it and how it is probably used.